allennlp/training/initializers.py

```diff
@@ -0,0 +1,117 @@
+
+        Parameters
+        ----------
+        module : torch.nn.Module, required.
+            The Pytorch module to apply the initializers to.
+        """
+        # Store which initialisers were applied to which parameters.
+        not_explicitly_initialized_parameters = []
+        for name, parameter in module.named_parameters():
+            is_initialized = False
+            for initializer_regex, initializer in self._initializers.items():
+                if self.parameter_name_matches_regex(initializer_regex, name):
+                    initializer(parameter)
+                    logger.info("Initializing %s using %s "
+                                "Intitializer.", name, initializer_regex)
+                    is_initialized = True
+            if not is_initialized:
+                not_explicitly_initialized_parameters.append((name, parameter))
+
+        for name, parameter in not_explicitly_initialized_parameters:
+            self._default_initializer(parameter)
+            logger.info("Initializing %s using the Default "
+                        "Intitializer. (Normal(0, 1) unless user specified.)", name)
+
+    @staticmethod
+    def parameter_name_matches_regex(regex: str, parameter_name: str):
+        return re.search(regex, parameter_name) is not None
+
+    @classmethod
+    def from_params(cls, params: Params) -> "InitializerApplicator":
+        """
+        Converts a Params object into an InitializerApplicator. The json should
+        be formatted as follows:
+
+        initializers: {
+            parameter_regex_match1: {
+                "type": "normal"
+                "mean": 0.01
+                "std": 0.1
+            },
+            parameter_regex_match2: "uniform",
+
+            "default": "orthogonal"
+        }
+        where the keys are regex matches to the parameters (with the exception of the "default" key,
+        which will be used as the default initializer for parameters which do not match any
+        initializer regex passed to the InitializerApplicator). The values can either be strings,
+        in which case they correspond to the names of initializers, or dictionaries, in which
+        case they must contain the "type" key, corresponding to the name of an initializer.
+        In addition, they may contain auxiliary named parameters which will be fed to the
+        initializer itself. To determine valid auxiliary parameters, please refer to the
+        torch.nn.init documentation.
+
+        Parameters
+        ----------
+        params: Params, required.
+            A Params object containing an "initializers" key.
+
+        Returns
+        -------
+        An InitializerApplicator containing the specified initializers.
+        """
+        # Construct a dictionary of available initializers from the torch.nn.init package.
+        torch_initalizer_names = [func for func in dir(torch.nn.init)
```

Comment:
Wow, does torch really not have a list of available initializers?  Same as keras...  Why don't people think this is a useful thing to have?  I don't get it.